synapse/third_party_rules/access_rules.py

```diff
@@ -394,6 +405,130 @@ async def check_visibility_can_be_modified(
+        if (
+            event.type == EventTypes.Member
+            and event.membership == Membership.LEAVE
+            and event.sender == event.state_key
+        ):
+            return True
+
+        if event.type == EventTypes.PowerLevels:
+            # Check if the power level event is associated with a room unfreeze (because
+            # the power level events will be sent before the frozen state event). This
+            # means we check that the users_default is back to 0 and the sender set
+            # themselves as admin.
+            current_power_levels = state_events.get((EventTypes.PowerLevels, ""))
+            if current_power_levels:
+                old_content = current_power_levels.content
+                old_content["users_default"] = 0
+
+                new_content = unfreeze(event.content)
+                sender_pl = new_content.get("users", {}).get(event.sender, 0)
+
+                # We don't care about the users section as long as the new event gives
+                # full power to the sender.
+                del old_content["users"]
+                del new_content["users"]
+
+                if new_content == old_content and sender_pl == 100:
+                    return True
+
+        return False
+
+    async def _on_frozen_state_change(
+        self,
+        event: EventBase,
+        state_events: StateMap[EventBase],
+    ) -> Union[bool, dict]:
+        frozen = event.content.get("frozen", None)
+        if not isinstance(frozen, bool):
+            # Invalid event: frozen is either missing or not a boolean.
+            return False
+
+        # If the event was sent from a restricted homeserver, don't allow the state
+        # change.
+        if (
+            UserID.from_string(event.sender).domain
+            in self.domains_forbidden_when_restricted
+        ):
+            return False
+
+        current_frozen_state = state_events.get(
+            (FROZEN_STATE_TYPE, ""),
+        )  # type: EventBase
+
+        if (
+            current_frozen_state is not None
+            and current_frozen_state.content.get("frozen") == frozen
+        ):
+            # This is a noop, accept the new event but don't do anything more.
+            return True
+
+        # If the event was received over federation, we want to accept it but not to
+        # change the power levels.
+        if not self._is_local_user(event.sender):
+            return True
+
+        current_power_levels = state_events.get(
+            (EventTypes.PowerLevels, ""),
+        )  # type: EventBase
+
+        power_levels_content = unfreeze(current_power_levels.content)
+
+        if not frozen:
+            # We're unfreezing the room: enforce the right value for the power levels so
+            # the room isn't in a weird/broken state afterwards.
+            users = power_levels_content.get("users", {})
+            users[event.sender] = 100
+            power_levels_content["users"] = users
```

Comment:
```suggestion
            users = power_levels_content.setdefault("users", {})
            users[event.sender] = 100
```

Reply:
Oooh I didn't know (or more likely forgot) `users` was a reference here, thanks!

Reply:
it's not the most critical thing, but it would be nice to clean this up.